pkg/controller/azvolume.go

```diff
@@ -81,260 +82,303 @@ func (r *ReconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Req
+		return nil
 	}
 
-	updated := azVolume.DeepCopy()
-	if updated.Status.State == state {
-		return azVolume, nil
+	if azVolume.ObjectMeta.Finalizers == nil {
+		return azVolume
 	}
-	updated.Status.State = state
 
-	if err := r.client.Update(ctx, updated, &client.UpdateOptions{}); err != nil {
-		klog.Errorf("failed to update AzVolume (%s): %v", volumeName, err)
-		return nil, err
+	finalizers := []string{}
+	for _, finalizer := range azVolume.ObjectMeta.Finalizers {
+		if exists := finalizersToDelete[finalizer]; exists {
+			continue
+		}
+		finalizers = append(finalizers, finalizer)
 	}
-	return updated, nil
+	azVolume.ObjectMeta.Finalizers = finalizers
+	return azVolume
 }
 
-func (r *ReconcileAzVolume) updateStatus(ctx context.Context, volumeName string, azVolume *v1alpha1.AzVolume, phase v1alpha1.AzVolumePhase, isDeleted bool, status *v1alpha1.AzVolumeStatusParams) (*v1alpha1.AzVolume, error) {
-	var err error
+func (r *ReconcileAzVolume) updateState(ctx context.Context, azVolume *v1alpha1.AzVolume, state v1alpha1.AzVolumeState) *v1alpha1.AzVolume {
 	if azVolume == nil {
-		azVolume, err = azureutils.GetAzVolume(ctx, r.client, r.azVolumeClient, volumeName, r.namespace, true)
-		if err != nil {
-			klog.Errorf("failed to get AzVolume (%s): %v", volumeName, err)
-			return azVolume, err
-		}
+		return nil
 	}
-
-	if isDeleted {
-		if azVolume, err = r.deleteFinalizer(ctx, volumeName, azVolume, map[string]bool{azureutils.AzVolumeFinalizer: true}, true); err != nil {
-			klog.Errorf("failed to delete finalizer (%s) for AzVolume (%s): %v", azureutils.AzVolumeFinalizer, volumeName, err)
-			return nil, err
-		}
-		return azVolume, nil
+	if azVolume.Status.State == state {
+		return azVolume
 	}
+	azVolume.Status.State = state
 
-	updated := azVolume.DeepCopy()
-	if updated.Status.Detail == nil {
-		if status != nil {
-			updated.Status.Detail = &v1alpha1.AzVolumeStatusDetail{
-				ResponseObject: status,
-			}
+	return azVolume
+}
+
+func (r *ReconcileAzVolume) updateStatusDetail(ctx context.Context, azVolume *v1alpha1.AzVolume, status *v1alpha1.AzVolumeStatusParams) *v1alpha1.AzVolume {
+	if azVolume == nil {
+		return nil
+	}
+	if azVolume.Status.Detail == nil {
+		azVolume.Status.Detail = &v1alpha1.AzVolumeStatusDetail{
+			ResponseObject: status,
 		}
 	} else {
-		// Updating status after update operation
 		if status != nil {
-			if updated.Status.Detail.ResponseObject == nil {
-				updated.Status.Detail.ResponseObject = &v1alpha1.AzVolumeStatusParams{}
+			if azVolume.Status.Detail.ResponseObject == nil {
+				azVolume.Status.Detail.ResponseObject = &v1alpha1.AzVolumeStatusParams{}
 			}
-			updated.Status.Detail.ResponseObject.CapacityBytes = status.CapacityBytes
-			updated.Status.Detail.ResponseObject.NodeExpansionRequired = status.NodeExpansionRequired
+			azVolume.Status.Detail.ResponseObject.CapacityBytes = status.CapacityBytes
+			azVolume.Status.Detail.ResponseObject.NodeExpansionRequired = status.NodeExpansionRequired
```

Comment:
We cannot fully overwrite `azVolume.Status.Detail.ResponseObject` with `status` as `cloudProvisioner.ExpandVolume` returns a partially filled `AzVolumeStatusParams` as below: 

https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/3379024d48f4ca67c687b3c32f7b20679f2c5aae/pkg/provisioner/cloudprovisioner.go#L489-L492

Reply:
Ah, I see. Can `status` ever be `nil`? If not, then it would probably be clearer to take `CapacityBytes` and `NodeExpansionRequired` as separate parameters to this function. Otherwise, a comment describing why we only take the subset would be helpful.

Reply:
Not that I can think of. It is expected that a call to this function input a non-nil `status` so I agree that taking `CapacityBytes` and `NodeExpansionRequired` as separate parameters or adding dedicated comment would make things clearer here.